SF iOS/SF iOS/Settings/SettingsViewController.m

```diff
@@ -0,0 +1,159 @@
+//
+//  SettingsViewController.m
+//  SF iOS
+//
+//  Created by Zachary Drayer on 4/7/19.
+//  Copyright © 2019 Amit Jain. All rights reserved.
+//
+
+#import "SettingsViewController.h"
+#import "NSUserDefaults+Settings.h"
+
+typedef NS_ENUM(NSInteger, SettingsSection) {
+    SettingsSectionDirectionsProvider,
+    SettingsSectionCount
+};
+
+typedef NS_ENUM(NSInteger, SettingsSectionDirectionsProviderValues) {
+    SettingsSectionDirectionsProviderApple,
+    SettingsSectionDirectionsProviderGoogle,
+    SettingsSectionDirectionsProviderCount
+};
+
+@interface SettingsViewController () <UITableViewDataSource, UITableViewDelegate>
+
+@property (nonatomic) UITableView *tableView;
+
+@end
+
+@implementation SettingsViewController
+
+- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
+    self = [super initWithNibName:nil bundle:nil];
+
+    _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
+    _tableView.dataSource = self;
+    _tableView.delegate = self;
+    [_tableView registerClass:UITableViewCell.class forCellReuseIdentifier:NSStringFromClass(UITableViewCell.class)];
+    _tableView.rowHeight = 64;
+    _tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
+    _tableView.contentInset = UIEdgeInsetsMake(40, 0, 0, 0);
+    _tableView.translatesAutoresizingMaskIntoConstraints = false;
+    _tableView.delaysContentTouches = NO;
+    _tableView.tintColor = UIColor.blackColor;
+
+    return self;
+}
+
+- (void)loadView {
+    [super loadView];
+
+    [self.view addSubview:self.tableView];
+
+    [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = true;
+    [self.tableView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = true;
+    [self.tableView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = true;
+    [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = true;
+    [self.tableView.widthAnchor constraintEqualToAnchor:self.view.widthAnchor].active = true;
+}
```

Comment:
This is a little different than the rest of the app the way you use loadView. I might try changing that others to use your way. It is a bit nicer imho

Reply:
yeah, its my preferred approach to separating out init from configuration. arguably, `viewDidLoad` could also work for setting the constraints, but it seemed silly to override it just for this ¯\\_(ツ)_/¯

Reply:
Also worth noting your use of _tableView in the initializer.

Reply:
Wow @zadr this is extremely good work. 
Question for learning purposes, what's significant about the use of `_tableView` in the initializer?  @rodericj

Reply:
There is technically a difference between calling `[_tableview someMethod]` vs `[self.tableView someMethod]`. From what I remember, calling `self.` calls a method that is auto generated through the `@property ....` in the interface. The method would be something like 

`- (UITableView *)tableView {
   return _tableview;
}
`

^Hey that's pretty helpful right? ^


That method can be overridden in your class....or more importantly, given that Objective-C is a HIGHLY dynamic language, this can be overridden by basically anyone anywhere in the app so at build time you have no guarantee that calling `self.tableView` isn't running this method:

```
- (UITableView *)tableView {
   [GlobalAnalyticsProvider takeSnapshotOfUserRightNowAndSendItToABadPlayer];
   return _tableView;
}
```


So it's a safety thing. Before Zach's addition here I was being slightly reckless if I'm being honest. So using _tableView becomes a policy decision.

That said, there may be other reasons, but also I may be entirely wrong. I reserve the right to be wrong.

Reply:
Basically!

This is a bit of defensive code. The idea is that during initialization, we don’t yet know what `self` is yet.

Because self can be re-assigned, and because we don’t what any getters will do— maybe there’s a subclass somewhere we don’t know about— we should avoid doing anything that can cause references to `self` to be sent out— by triggering KVO notifications, posting notifications, or any other form of calling out to other code— before our object is fully initialized.

So, to avoid triggering these kinds of behavior, it’s something of a best practice to prefer _instanceVariable access in init and dealloc, and .property access everywhere else with ObjC.